contributors/design-proposals/storage/csi-migration.md

```diff
@@ -224,78 +284,300 @@ the same ways as we are not removing/adding PV’s or PVC’s in out of the ordi
+	// that driver.
+	Name string `json:"name"`
+
+	// available is a boolean representing whether the driver has been installed
+	// on this node or not.
+	Available bool `json:"available"`
+
+	// volumePluginMechanism announces what mechanism underlies volume plugins.
+	// It is set by Kubelet. It is used by the attach/detach controller, which
+	// needs to know how to perform attachments. The allowed values are:
+	// * "in-tree": the volume operation (e.g., attach/detach) ought to be
+	//   directly performed by the attach/detach controller.
+	// * "csi-plugin": the attach/detach controller ought to request
+	//   the csi plugin to perform the volume operation rather than perform it directly.
+	VolumePluginMechanism VolumePluginMechanism `json:"volumePluginMechanism"`
+}
+
+// CSIDriverInfoSpec holds information about the specification of one CSI driver installed on a node
+type CSIDriverInfoSpec struct {
+	// name is the name of the CSI driver that this object refers to.
+	// This MUST be the same name returned by the CSI GetPluginName() call for
+	// that driver.
+	Name string `json:"name"`
+
+	// nodeID of the node from the driver point of view.
+	// This field enables Kubernetes to communicate with storage systems that do
+	// not share the same nomenclature for nodes. For example, Kubernetes may
+	// refer to a given node as "node1", but the storage system may refer to
+	// the same node as "nodeA". When Kubernetes issues a command to the storage
+	// system to attach a volume to a specific node, it can use this field to
+	// refer to the node name using the ID that the storage system will
+	// understand, e.g. "nodeA" instead of "node1".
+	NodeID string `json:"nodeID"`
+
+	// topologyKeys is the list of keys supported by the driver.
+	// When a driver is initialized on a cluster, it provides a set of topology
+	// keys that it understands (e.g. "company.com/zone", "company.com/region").
+	// When a driver is initialized on a node it provides the same topology keys
+	// along with values that kubelet applies to the coresponding node API
+	// object as labels.
+	// When Kubernetes does topology aware provisioning, it can use this list to
+	// determine which labels it should retrieve from the node object and pass
+	// back to the driver.
+	TopologyKeys []string `json:"topologyKeys"`
+}
+```
+
+#### Spec/Status Split
+
+The new `CSINodeInfo` API has been split up to have a `Spec` and a `Status`. The
+reason being that there are some fields that are constant after being set by the
+CSI Driver on initialization: `NodeID` and `TopologyKeys`. Whereas there are
+some other fields that are subject to change: `Available` and
+`VolumePluginMechanism`. There are future plans to reconcile the `Status` based
+on periodically polling the driver.
+
+Because of this split there is a possibility that there will be a `Spec` with no
+corresponding `Status` or a `Status` without a corresponding `Spec`. All
```

Comment:
Can you really have a Status without a Spec? How would that happen?

Reply:
I have no idea, but that's why its an undefined state.